File: app/src/androidTest/java/org/fdroid/fdroid/LocalizationTest.java
package org.fdroid.fdroid;

import android.app.Instrumentation;
import android.content.Context;
import android.content.res.AssetManager;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.util.Log;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.HashSet;
import java.util.IllegalFormatException;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Runs through all of the translated strings and tests them with the same format
 * values that the source strings expect.  This is to ensure that the formats in
 * the translations are correct in number and in type (e.g. {@code s} or {@code s}.
 * It reads the source formats and then builds {@code formats} to represent the
 * position and type of the formats.  Then it runs through all of the translations
 * with formats of the correct number and type.
 */
@RunWith(AndroidJUnit4.class)
public class LocalizationTest {
    public static final String TAG = "LocalizationTest";

    private final Pattern androidFormat = Pattern.compile("(%[a-z0-9]\\$?[a-z]?)");
    private final Locale[] locales = Locale.getAvailableLocales();
    private final HashSet<String> localeNames = new HashSet<>(locales.length);

    private AssetManager assets;
    private Configuration config;
    private Resources resources;

    @Before
    public void setUp() {
        for (Locale locale : Languages.LOCALES_TO_TEST) {
            localeNames.add(locale.toString());
        }
        for (Locale locale : locales) {
            localeNames.add(locale.toString());
        }

        Instrumentation instrumentation = InstrumentationRegistry.getInstrumentation();
        Context context = instrumentation.getTargetContext();
        assets = context.getAssets();
        config = context.getResources().getConfiguration();
        config.locale = Locale.ENGLISH;
        // Resources() requires DisplayMetrics, but they are only needed for drawables
        resources = new Resources(assets, new DisplayMetrics(), config);
    }

    @Test
    public void testLoadAllPlural() throws IllegalAccessException {
        Field[] fields = R.plurals.class.getDeclaredFields();

        HashMap<String, String> haveFormats = new HashMap<>();
        for (Field field : fields) {
            //Log.i(TAG, field.getName());
            int resId = field.getInt(int.class);
            CharSequence string = resources.getQuantityText(resId, 4);
            //Log.i(TAG, field.getName() + ": '" + string + "'");
            Matcher matcher = androidFormat.matcher(string);
            int matches = 0;
            char[] formats = new char[5];
            while (matcher.find()) {
                String match = matcher.group(0);
                char formatType = match.charAt(match.length() - 1);
                switch (match.length()) {
                    case 2:
                        formats[matches] = formatType;
                        matches++;
                        break;
                    case 4:
                        formats[Integer.parseInt(match.substring(1, 2)) - 1] = formatType;
                        break;
                    case 5:
                        formats[Integer.parseInt(match.substring(1, 3)) - 1] = formatType;
                        break;
                    default:
                        throw new IllegalStateException(field.getName() + " has bad format: " + match);
                }
            }
            haveFormats.put(field.getName(), new String(formats).trim());
        }

        for (Locale locale : locales) {
            config.locale = locale;
            // Resources() requires DisplayMetrics, but they are only needed for drawables
            resources = new Resources(assets, new DisplayMetrics(), config);
            for (Field field : fields) {
                String formats = null;
                try {
                    int resId = field.getInt(int.class);
                    for (int quantity = 0; quantity < 567; quantity++) {
                        resources.getQuantityString(resId, quantity);
                    }

                    formats = haveFormats.get(field.getName());
                    switch (formats) {
                        case "d":
                            resources.getQuantityString(resId, 1, 1);
                            break;
                        case "s":
                            resources.getQuantityString(resId, 1, "ONE");
                            break;
                        case "ds":
                            resources.getQuantityString(resId, 2, 1, "TWO");
                            break;
                        default:
                            if (!TextUtils.isEmpty(formats)) {
                                throw new IllegalStateException("Pattern not included in tests: " + formats);
                            }
                    }
                } catch (IllegalFormatException | Resources.NotFoundException e) {
                    Log.i(TAG, locale + " " + field.getName());
                    throw new IllegalArgumentException("Bad '" + formats + "' format in " + locale + " "
                            + field.getName() + ": " + e.getMessage());
                }
            }
        }
    }

    @Test
    public void testLoadAllStrings() throws IllegalAccessException {
        Field[] fields = R.string.class.getDeclaredFields();

        HashMap<String, String> haveFormats = new HashMap<>();
        for (Field field : fields) {
            String string = resources.getString(field.getInt(int.class));
            Matcher matcher = androidFormat.matcher(string);
            int matches = 0;
            char[] formats = new char[5];
            while (matcher.find()) {
                String match = matcher.group(0);
                char formatType = match.charAt(match.length() - 1);
                switch (match.length()) {
                    case 2:
                        formats[matches] = formatType;
                        matches++;
                        break;
                    case 4:
                        formats[Integer.parseInt(match.substring(1, 2)) - 1] = formatType;
                        break;
                    case 5:
                        formats[Integer.parseInt(match.substring(1, 3)) - 1] = formatType;
                        break;
                    default:
                        throw new IllegalStateException(field.getName() + " has bad format: " + match);
                }
            }
            haveFormats.put(field.getName(), new String(formats).trim());
        }

        for (Locale locale : locales) {
            config.locale = locale;
            // Resources() requires DisplayMetrics, but they are only needed for drawables
            resources = new Resources(assets, new DisplayMetrics(), config);
            for (Field field : fields) {
                int resId = field.getInt(int.class);
                resources.getString(resId);

                String formats = haveFormats.get(field.getName());
                try {
                    switch (formats) {
                        case "d":
                            resources.getString(resId, 1);
                            break;
                        case "dd":
                            resources.getString(resId, 1, 2);
                            break;
                        case "dds":
                            resources.getString(resId, 1, 2, "THREE");
                            break;
                        case "sds":
                            resources.getString(resId, "ONE", 2, "THREE");
                            break;
                        case "s":
                            resources.getString(resId, "ONE");
                            break;
                        case "ss":
                            resources.getString(resId, "ONE", "TWO");
                            break;
                        case "sss":
                            resources.getString(resId, "ONE", "TWO", "THREE");
                            break;
                        case "ssss":
                            resources.getString(resId, "ONE", "TWO", "THREE", "FOUR");
                            break;
                        case "ssd":
                            resources.getString(resId, "ONE", "TWO", 3);
                            break;
                        case "sssd":
                            resources.getString(resId, "ONE", "TWO", "THREE", 4);
                            break;
                        default:
                            if (!TextUtils.isEmpty(formats)) {
                                throw new IllegalStateException("Pattern not included in tests: " + formats);
                            }
                    }
                } catch (Exception e) {
                    Log.i(TAG, locale + " " + field.getName());
                    throw new IllegalArgumentException("Bad format in '" + locale + "' '" + field.getName() + "': "
                            + e.getMessage());
                }
            }
        }
    }
}

File: app/src/androidTest/java/org/fdroid/ui/LocalizationTest.kt
package org.fdroid.ui

import android.content.res.Resources
import android.text.TextUtils
import android.util.DisplayMetrics
import android.util.Log
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import java.util.IllegalFormatException
import java.util.Locale
import java.util.regex.Pattern
import org.fdroid.R
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Runs through all the translated strings and tests them with the same format values that the
 * source strings expect. This is to ensure that the formats in the translations are correct in
 * number and in type (e.g. `s` or `s`). It reads the source formats and then builds `formats` to
 * represent the position and type of the formats. Then it runs through all the translations with
 * formats of the correct number and type.
 */
@Suppress("DEPRECATION")
@RunWith(AndroidJUnit4::class)
class LocalizationTest {
  private val androidFormat: Pattern = Pattern.compile("(%[a-z0-9]\\$?[a-z]?)")
  private val locales: Array<Locale> = Locale.getAvailableLocales()
  private val localeNames = HashSet<String?>(locales.size)

  private val context = InstrumentationRegistry.getInstrumentation().targetContext
  private var assets = context.assets
  private var config =
    context.resources.configuration.apply {
      locale = Locale.ENGLISH
    }
  private var resources = Resources(assets, DisplayMetrics(), config)

  @Before
  fun setUp() {
    for (locale in LOCALES_TO_TEST) {
      localeNames.add(locale.toString())
    }
    for (locale in locales) {
      localeNames.add(locale.toString())
    }
  }

  @Test
  @Throws(IllegalAccessException::class)
  fun testLoadAllPlural() {
    val fields = R.plurals::class.java.declaredFields

    val haveFormats = HashMap<String?, String?>()
    for (field in fields) {
      // Log.i(TAG, field.getName());
      val resId = field.getInt(Int::class.javaPrimitiveType)
      val string = resources.getQuantityText(resId, 4)
      // Log.i(TAG, field.getName() + ": '" + string + "'");
      val matcher = androidFormat.matcher(string)
      var matches = 0
      val formats = CharArray(5)
      while (matcher.find()) {
        val match = matcher.group(0)
        val formatType = match!!.get(match.length - 1)
        when (match.length) {
          2 -> {
            formats[matches] = formatType
            matches++
          }
          4 -> formats[match.substring(1, 2).toInt() - 1] = formatType
          5 -> formats[match.substring(1, 3).toInt() - 1] = formatType
          else -> throw IllegalStateException(field.getName() + " has bad format: " + match)
        }
      }
      haveFormats[field.getName()] = String(formats).trim { it <= ' ' }
    }

    for (locale in locales) {
      config.locale = locale
      // Resources() requires DisplayMetrics, but they are only needed for drawables
      resources = Resources(assets, DisplayMetrics(), config)
      for (field in fields) {
        var formats: String? = null
        try {
          val resId = field.getInt(Int::class.javaPrimitiveType)
          for (quantity in 0..566) {
            resources.getQuantityString(resId, quantity)
          }

          formats = haveFormats[field.getName()]
          when (formats) {
            "d" -> resources.getQuantityString(resId, 1, 1)
            "s" -> resources.getQuantityString(resId, 1, "ONE")
            "ds" -> resources.getQuantityString(resId, 2, 1, "TWO")
            else ->
              check(TextUtils.isEmpty(formats)) { "Pattern not included in tests: " + formats }
          }
        } catch (e: IllegalFormatException) {
          Log.i(TAG, locale.toString() + " " + field.getName())
          throw IllegalArgumentException(
            ("Bad '" + formats + "' format in " + locale + " " + field.getName() + ": " + e.message)
          )
        } catch (e: Resources.NotFoundException) {
          Log.i(TAG, locale.toString() + " " + field.getName())
          throw IllegalArgumentException(
            ("Bad '" + formats + "' format in " + locale + " " + field.getName() + ": " + e.message)
          )
        }
      }
    }
  }

  @Test
  @Throws(IllegalAccessException::class)
  fun testLoadAllStrings() {
    val fields = R.string::class.java.declaredFields

    val haveFormats = HashMap<String?, String?>()
    for (field in fields) {
      val string = resources.getString(field.getInt(Int::class.javaPrimitiveType))
      val matcher = androidFormat.matcher(string)
      var matches = 0
      val formats = CharArray(5)
      while (matcher.find()) {
        val match = matcher.group(0)
        val formatType = match!!.get(match.length - 1)
        when (match.length) {
          2 -> {
            formats[matches] = formatType
            matches++
          }
          4 -> formats[match.substring(1, 2).toInt() - 1] = formatType
          5 -> formats[match.substring(1, 3).toInt() - 1] = formatType
          else -> throw IllegalStateException(field.getName() + " has bad format: " + match)
        }
      }
      haveFormats[field.getName()] = String(formats).trim { it <= ' ' }
    }

    for (locale in locales) {
      config!!.locale = locale
      // Resources() requires DisplayMetrics, but they are only needed for drawables
      resources = Resources(assets, DisplayMetrics(), config)
      for (field in fields) {
        val resId = field.getInt(Int::class.javaPrimitiveType)
        resources.getString(resId)

        val formats = haveFormats.get(field.getName())
        try {
          when (formats) {
            "d" -> resources.getString(resId, 1)
            "dd" -> resources.getString(resId, 1, 2)
            "ds" -> resources.getString(resId, 1, "TWO")
            "dds" -> resources.getString(resId, 1, 2, "THREE")
            "sds" -> resources.getString(resId, "ONE", 2, "THREE")
            "s" -> resources.getString(resId, "ONE")
            "ss" -> resources.getString(resId, "ONE", "TWO")
            "sss" -> resources.getString(resId, "ONE", "TWO", "THREE")
            "ssss" -> resources.getString(resId, "ONE", "TWO", "THREE", "FOUR")
            "ssd" -> resources.getString(resId, "ONE", "TWO", 3)
            "sssd" -> resources.getString(resId, "ONE", "TWO", "THREE", 4)
            else ->
              check(TextUtils.isEmpty(formats)) { "Pattern not included in tests: " + formats }
          }
        } catch (e: Exception) {
          Log.i(TAG, locale.toString() + " " + field.getName())
          throw IllegalArgumentException(
            ("Bad format in '" + locale + "' '" + field.getName() + "': " + e.message)
          )
        }
      }
    }
  }

  companion object {
    const val TAG: String = "LocalizationTest"

    val LOCALES_TO_TEST: Array<Locale> =
      arrayOf(
        Locale.ENGLISH,
        Locale.FRENCH,
        Locale.GERMAN,
        Locale.ITALIAN,
        Locale.JAPANESE,
        Locale.KOREAN,
        Locale.SIMPLIFIED_CHINESE,
        Locale.TRADITIONAL_CHINESE,
        Locale("zh", "HK"),
        Locale("bo"),
        Locale("af"),
        Locale("ar"),
        Locale("be"),
        Locale("bg"),
        Locale("ca"),
        Locale("cs"),
        Locale("da"),
        Locale("el"),
        Locale("es"),
        Locale("eo"),
        Locale("et"),
        Locale("eu"),
        Locale("fa"),
        Locale("fi"),
        Locale("he"),
        Locale("hi"),
        Locale("hu"),
        Locale("hy"),
        Locale("id"),
        Locale("is"),
        Locale("it"),
        Locale("ml"),
        Locale("my"),
        Locale("nb"),
        Locale("nl"),
        Locale("pl"),
        Locale("pt"),
        Locale("ro"),
        Locale("ru"),
        Locale("sc"),
        Locale("sk"),
        Locale("sn"),
        Locale("sr"),
        Locale("sv"),
        Locale("th"),
        Locale("tr"),
        Locale("uk"),
        Locale("vi"),
      )
  }
}
